### PR TITLE
Updating backend docstrings

### PR DESCRIPTION
**Pull request by Transurgeon**
## Description
This PR attempts to add docstrings to all functions in the canonicalization backend. Notably, this includes the new NumPy backend that has recently been added in #2186 . This branch is WIP but a good chunk has already been completed, so a first review isn't out of the question. I also found a neat way to refactor indexing in the kron functions. 

P.S: Adding non-trivial docstrings is actually kind of challenging, please bare with all the redundancy and reiterations that I may have introduced.

## Type of change
- [ ] New feature (backwards compatible)
- [ ] New feature (breaking API changes)
- [ ] Bug fix
- [x] Other (Documentation, CI, ...)

## [Contribution checklist](https://www.cvxpy.org/contributing/index.html#contribution-checklist)
- [x] Add our license to new files.
- [x] Check that your code adheres to our coding style.
- [x] Write unittests.
- [x] Run the unittests and check that they’re passing.
- [x] Run the benchmarks to make sure your change doesn’t introduce a regression.

### CODE REVIEW

**Comment by SteveDiamond** [comment]
Great work adding these docstrings! Having docstrings sets a good pattern for future developers.

File: cvxpy/lin_ops/canon_backend.py [hunk 1]
```diff
@@ -509,11 +509,24 @@ def kron_l(self, lin: LinOp, view: TensorView) -> TensorView:
...
+            .flatten(order="F").astype(int)
+        return row_indices
+
     @abstractmethod
     def get_variable_tensor(self, shape: tuple[int, ...], variable_id: int) -> Any:
         """
-        Returns tensor of a variable node, i.e., eye(n) across axes 0 and 1, where n i the number of
-        entries of the variable.
+        Returns tensor of a variable node, i.e., eye(n) across axes 0 and 1, where n is
+        the number of entries of the variable.
```
Comment (phschiele) [inline cvxpy/lin_ops/canon_backend.py:529]:
```suggestion
        Returns tensor of a variable node, i.e., eye(n) across axes 0 and 1, where n is
        the size of the variable.
```

File: cvxpy/lin_ops/canon_backend.py [hunk 2]
```diff
@@ -527,13 +540,18 @@ def get_data_tensor(self, data: Any) -> Any:
     @abstractmethod
     def get_param_tensor(self, shape: tuple[int, ...], parameter_id: int) -> Any:
         """
-        Returns tensor of a parameter node, i.e., eye(n) across axes 0 and 2, where n i the number
-        of entries of the parameter.
+        Returns tensor of a parameter node, i.e., eye(n) across axes 0 and 2, where n is
+        the number of entries of the parameter.
```
Comment (phschiele) [inline cvxpy/lin_ops/canon_backend.py:544]:
```suggestion
        Returns tensor of a parameter node, i.e., eye(n) across axes 0 and 2, where n is
        the size of the parameter.
```

File: cvxpy/lin_ops/canon_backend.py [hunk 3]
```diff
@@ -1029,25 +1178,57 @@ def kron_r(self, lin: LinOp, view: NumpyTensorView) -> NumpyTensorView:
...
+        [[[1   0   0   0],
+         [0   1   0   0],
+         [2   0   0   0],
+         [0   2   0   0],
+         [0   0   1   0],
+         [0   0   0   1],
+         [0   0   2   0],
+         [0   0   0   2]]]
+
+         However computing kron(x, a) directly gives us:
```
Comment (Transurgeon) [inline cvxpy/lin_ops/canon_backend.py:1219]: I added a comparison of normal kron(x, a) to show why reordering is necessary. I am not sure if this is a great idea as it makes the docstring very lengthy.

File: cvxpy/lin_ops/canon_backend.py [hunk 4]
```diff
@@ -764,25 +847,22 @@ def kron_r(self, lin: LinOp, view: ScipyTensorView) -> ScipyTensorView:
...
-        lhs_ones = np.ones(lin.data.shape)
-        rhs_ones = np.ones(rhs_shape)
-
-        lhs_arange = np.arange(np.prod(lin.data.shape)).reshape(lin.data.shape, order="F")
-        rhs_arange = np.arange(np.prod(rhs_shape)).reshape(rhs_shape, order="F")
-
-        row_indices = (np.kron(lhs_ones, rhs_arange) +
-                       np.kron(lhs_arange, rhs_ones * np.prod(rhs_shape))) \
-            .flatten(order="F").astype(int)
+        # TODO: add test to verify that reordering is necessary
```
Comment (Transurgeon) [inline cvxpy/lin_ops/canon_backend.py:850]: Added TODO for the kron_r unit-test.

File: cvxpy/lin_ops/canon_backend.py (hunk 3 above)
Comment (phschiele) [reply cvxpy/lin_ops/canon_backend.py:1237]: I think we should move this to the tests and add a pointer here (which we should do in general).

Comment (Transurgeon) [reply cvxpy/lin_ops/canon_backend.py:1237]: Ok, I'll try to make some additional changes now.

**Review by phschiele — APPROVED** [review]
Good from my side.

**Comment by phschiele** [comment]
Benchmarks are failing because of a new asv release. Merging here as it only added docs.

**Pull request merged**